pkg/tls/config.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tls
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// TODO
+type WatchedConfig struct {
+	log         logrus.FieldLogger
+	caFiles     []string
+	certFile    string
+	privkeyFile string
+	stop        chan struct{}
+	mu          lock.Mutex
+	caCertPool  *x509.CertPool
+	keypair     *tls.Certificate
+}
+
+var (
+	CertWithoutPrivkeyErr = errors.New("certificate and private key are both required, but only the certificate was provided")
+	PrivkeyWithoutCertErr = errors.New("certificate and private key are both required, but only the private key was provided")
+)
+
+// TODO
+func NewWatchedConfig(log logrus.FieldLogger, caFiles []string, certFile, privkeyFile string) (*WatchedConfig, error) {
+	if certFile != "" && privkeyFile == "" {
+		return nil, CertWithoutPrivkeyErr
+	}
+	if certFile == "" && privkeyFile != "" {
+		return nil, PrivkeyWithoutCertErr
+	}
```

Comment:
The way I understand this is that the caller needs to provide both `certFile` and `privkeyFile` or neither. The latter case only makes sense if the caller provides `caFiles` in order to watch for CA files only. However, if the caller doesn't provide any of `caFiles`, `certFile` or `privkeyFile`, then it's not an error although the `WatchedConfig` is virtually useless. Do we want to check for this case as well?

Reply:
It can still be used to get a `*tls.Config` for TLS client using the system CAs. You are correct that the watching part will be useless though.

Reply:
In that case, it might make sense to rename `start()` to `Watch()` and remove the call from the constructor. This way, the config could also be used without active watching (and we should probably then rename `WatchedConfig` to `Config`).

Continuing on that topic, `update()` could possibly be made public as well (eg `Reload()`) so that users could create a `Config` and say call `Reload()` upon receiving a SIGHUP (and in this case, they wouldn't call `Watch()`).